modules/module-mysql/src/replication/zongji/BinLogListener.ts

```diff
@@ -67,126 +107,164 @@ export class BinLogListener {
+  }
 
-      this.zongji.on('stopped', () => {
+  private async stopZongji(): Promise<void> {
+    await new Promise<void>((resolve) => {
+      this.zongji.once('stopped', () => {
         resolve();
-        this.logger.info('BinLog listener stopped. Replication ended.');
       });
+      this.zongji.stop();
     });
+
+    // Wait until all the current events in the processing queue are also processed
+    if (this.processingQueue.length() > 0) {
+      await this.processingQueue.empty();
+    }
   }
 
-  public stop(): void {
-    if (!this.isStopped) {
-      this.zongji.stop();
-      this.processingQueue.kill();
+  public async stop(): Promise<void> {
+    if (!(this.isStopped || this.isStopping)) {
+      this.logger.info('Stopping BinLog Listener...');
+      this.isStopping = true;
+      await new Promise<void>((resolve) => {
+        if (this.isStopped) {
+          resolve();
+        }
+        this.zongji.once('stopped', () => {
+          this.isStopped = true;
+          this.logger.info('BinLog Listener stopped. Replication ended.');
+          resolve();
+        });
+        this.zongji.stop();
+        this.processingQueue.kill();
+      });
     }
   }
 
-  private get isStopped(): boolean {
-    return this.zongji.stopped;
+  public async replicateUntilStopped(): Promise<void> {
+    while (!this.isStopped) {
+      await timers.setTimeout(1_000);
+    }
+  }
+
+  private createProcessingQueue(): async.QueueObject<BinLogEvent> {
+    const queue = async.queue(this.createQueueWorker(), 1);
+
+    queue.error((error) => {
+      if (!(this.isStopped || this.isStopping)) {
+        this.logger.error('Error processing BinLog event:', error);
+        this.stop();
+      } else {
+        this.logger.warn('Error processing BinLog event during shutdown:', error);
+      }
+    });
+
+    return queue;
   }
 
   private createZongjiListener(): ZongJi {
     const zongji = this.connectionManager.createBinlogListener();
 
     zongji.on('binlog', async (evt) => {
-      this.logger.info(`Received Binlog event:${evt.getEventName()}`);
-      this.processingQueue.push(evt);
-      this.queueMemoryUsage += evt.size;
-
-      // When the processing queue grows past the threshold, we pause the binlog listener
-      if (this.isQueueOverCapacity()) {
-        this.logger.info(
-          `Binlog processing queue has reached its memory limit of [${this.connectionManager.options.binlog_queue_memory_limit}MB]. Pausing Binlog listener.`
-        );
-        zongji.pause();
-        const resumeTimeoutPromise = new Promise((resolve) => {
-          setTimeout(() => resolve('timeout'), MAX_QUEUE_PAUSE_TIME_MS);
-        });
+      this.logger.debug(`Received BinLog event:${evt.getEventName()}`);
+      // We have to handle schema change events before handling more binlog events,
+      // This avoids a bunch of possible race conditions
+      if (zongji_utils.eventIsQuery(evt)) {
+        await this.processQueryEvent(evt);
```

Comment:
What specific race conditions are covered here?

It feels like this has the possibility to introduce more race conditions, since the order is not preserved between schema changes (applied immediately here) and data modifications (queued).

Furthermore, zongi doesn't wait for these promises to complete AFAIK, so processing here can cause concurrent processing of events if multiple events come in shortly after each other, and uncaught rejections if the processing fails.